sdk/azcore/internal/shared/response_error_test.go

```diff
@@ -0,0 +1,385 @@
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = NewResponseError(&http.Response{
+		Status:     "the system is down",
+		StatusCode: http.StatusInternalServerError,
+		Body:       io.NopCloser(strings.NewReader(`{ "error": { "code": "ErrorItsBroken", "message": "it's not working" } }`)),
+		Request: &http.Request{
+			Method: http.MethodGet,
+			URL:    fakeURL,
+		},
+	})
+	re, ok := err.(*ResponseError)
+	if !ok {
+		t.Fatalf("unexpected error type %T", err)
+	}
+	if c := re.StatusCode; c != http.StatusInternalServerError {
+		t.Fatalf("unexpected status code %d", c)
+	}
+	const want = `GET https://fakeurl.com/the/path
+--------------------------------------------------------------------------------
+RESPONSE 500: the system is down
+ERROR CODE: ErrorItsBroken
+--------------------------------------------------------------------------------
+{
+  "error": {
+    "code": "ErrorItsBroken",
+    "message": "it's not working"
+  }
+}
+--------------------------------------------------------------------------------
+`
+	if got := re.Error(); got != want {
+		t.Fatalf("\ngot:\n%s\nwant:\n%s\n", got, want)
+	}
+}
+
+func TestNewResponseErrorNoErrorCodeInvalidBody(t *testing.T) {
+	fakeURL, err := url.Parse("https://fakeurl.com/the/path?qp=removed")
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = NewResponseError(&http.Response{
+		Status:     "the system is down",
+		StatusCode: http.StatusInternalServerError,
+		Body:       io.NopCloser(strings.NewReader("JSON error string")),
+		Request: &http.Request{
+			Method: http.MethodGet,
+			URL:    fakeURL,
+		},
+	})
+	re, ok := err.(*ResponseError)
+	if !ok {
+		t.Fatalf("unexpected error type %T", err)
+	}
+	if c := re.StatusCode; c != http.StatusInternalServerError {
+		t.Fatalf("unexpected status code %d", c)
+	}
+	const want = `GET https://fakeurl.com/the/path
+--------------------------------------------------------------------------------
+RESPONSE 500: the system is down
+ERROR CODE UNAVAILABLE
+--------------------------------------------------------------------------------
+JSON error string
+--------------------------------------------------------------------------------
+`
+	if got := re.Error(); got != want {
+		t.Fatalf("\ngot:\n%s\nwant:\n%s\n", got, want)
+	}
+}
+
+type readFailer struct{}
+
+func (r *readFailer) Close() error {
+	return nil
+}
+
+func (r *readFailer) Read(p []byte) (int, error) {
+	return 0, errors.New("mock read failure")
+}
+
+func TestNewResponseErrorNoErrorCodeCantReadBody(t *testing.T) {
+	fakeURL, err := url.Parse("https://fakeurl.com/the/path?qp=removed")
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = NewResponseError(&http.Response{
+		Status:     "the system is down",
+		StatusCode: http.StatusInternalServerError,
+		Body:       &readFailer{},
+		Request: &http.Request{
+			Method: http.MethodGet,
+			URL:    fakeURL,
+		},
+	})
+	_, ok := err.(*ResponseError)
+	if ok {
+		t.Fatalf("unexpected error type %T", err)
+	}
```

Comment:
Are you expecting this `err` not to be converted to a `ResponseError`? Should line 271 read `if !ok {` ?

Reply:
It used to, but after discussing with Jeff yesterday it's changed.  In the case where reading from the HTTP response body fails, we return that error instead.

Reply:
The reason being, the read really shouldn't fail.  If it does, it's a bug so we don't want to mask it behind a `ResponseError`.